### PR TITLE
(SERVER-2526) Add basic testing for thread safety

### PR DESCRIPTION
**Pull request by pcarlisle**
Run with `lein thread-test` and then run ext/thread_test/test.rb. See
the file for setting iterations and number of processes. Individual
processes will stop on the first error and cause the main process to
exit 1 if there is a failure.

### CODE REVIEW

File: ext/thread_test/puppetserver.conf [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    master-var-dir: ${HOME}"/.puppetlabs/opt/puppet/cache"
+
+    # (optional) path to puppet run dir; if not specified, will use the puppet default
+    master-run-dir: ${HOME}"/.puppetlabs/var/run"
+
+    # (optional) path to puppet log dir; if not specified, will use the puppet default
+    master-log-dir: ${HOME}"/.puppetlabs/var/log"
+
+    # (optional) maximum number of JRuby instances to allow
+    max-active-instances: 1
```
Comment (Magisus) [inline ext/thread_test/puppetserver.conf]: With the ReferencePool, this setting is used to control the number of threads that can go through the instance at once. Once we have integrated that, we probably want to set it higher to properly test the multithreaded behavior.

Related tangential question: should we double up on the usage of this setting, or create new ones specific to the multithreaded environment?

File: ext/thread_test/Gemfile [hunk 2]
```diff
@@ -0,0 +1 @@
+gem 'rspec-expectations'
```
Comment (justinstoller) [inline ext/thread_test/Gemfile]: I needed to add a 
```rb
source 'https://rubygems.org'
```
in order fro this to work

Comment (justinstoller) [reply ext/thread_test/Gemfile]: Is there a reason we don't use the Gemfile at the root of the project?

**Comment by justinstoller** [comment]
👍 was able to cause threading issues with this

**Comment by Magisus** [comment]
I was also able successfully cause threading errors when I combined this with https://github.com/puppetlabs/puppetserver/pull/2236 and turned the `max-active-instances` up to 2. This uses the new JRuby pool architecture to achieve concurrent runs, rather than the interim implementation. Turning `max-active-instances` back down to 1 prevented the errors (as expected). In both cases the server was able to shut down gracefully following a ^C.

File: ext/thread_test/puppetserver.conf (hunk 1 above)
Comment (pcarlisle) [reply ext/thread_test/puppetserver.conf]: Honestly I'm not sure

Comment (Magisus) [reply ext/thread_test/puppetserver.conf]: We can leave the setting the same for now and update it if it seems bad or confusing once we show it to other people. I think it's okay if the API is a little fluid at the start.

**Pull request merged**